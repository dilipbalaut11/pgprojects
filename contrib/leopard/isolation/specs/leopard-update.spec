# Update tests for Leopard

setup
{
 CREATE TABLE ltest
 (   id  bigint not null primary key 
 ,   val bigint not null
 ) WITH (autovacuum_enabled = off);
 INSERT INTO ltest SELECT generate_series(1,182), 0;
}

teardown
{
  DROP TABLE ltest;
}

session s1
setup		{ SET enable_seqscan = off; }
step s1b	{ BEGIN TRANSACTION ISOLATION LEVEL REPEATABLE READ; SELECT count(*) FROM ltest; }
step s1s	{ SELECT * FROM ltest WHERE id IN (1,3) ORDER BY id; }
step s1c	{ COMMIT; }

session s2
setup		{ SET enable_seqscan = off; }
step s		{ SELECT * FROM ltest WHERE id IN (1,3) ORDER BY id; }
step u1		{ UPDATE ltest SET val = val + 1 WHERE id = 1; }
step u3		{ UPDATE ltest SET val = val + 1 WHERE id = 3; }
step srda	{ SELECT toid, root_tid, row_xmin, row_xmax, next_tid FROM rda.rda; }

# earlier updates have left space that can be cleared
permutation s1b u3 u3 s u1 s u1 s srda s1s u1 s s1c

# earlier updates in same chain may be cleared
permutation s1b u3 u3 s u3 s u1 s srda s1s u1 s s1c

permutation     u3 u3 s u3 s u3 s u3 s
