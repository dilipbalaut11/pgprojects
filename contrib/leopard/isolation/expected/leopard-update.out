Parsed test spec with 2 sessions

starting permutation: s1b u3 u3 s u1 s u1 s s1s u1 s s1c
step s1b: BEGIN TRANSACTION ISOLATION LEVEL REPEATABLE READ; SELECT count(*) FROM ltest;
count
-----
  182
(1 row)

step u3: UPDATE ltest SET val = val + 1 WHERE id = 3;
step u3: UPDATE ltest SET val = val + 1 WHERE id = 3;
step s: SELECT * FROM ltest WHERE id IN (1,3) ORDER BY id;
id|val
--+---
 1|  0
 3|  2
(2 rows)

step u1: UPDATE ltest SET val = val + 1 WHERE id = 1;
step s: SELECT * FROM ltest WHERE id IN (1,3) ORDER BY id;
id|val
--+---
 1|  1
 3|  2
(2 rows)

step u1: UPDATE ltest SET val = val + 1 WHERE id = 1;
step s: SELECT * FROM ltest WHERE id IN (1,3) ORDER BY id;
id|val
--+---
 1|  2
 3|  2
(2 rows)

step s1s: SELECT * FROM ltest WHERE id IN (1,3) ORDER BY id;
id|val
--+---
 1|  0
 3|  0
(2 rows)

step u1: UPDATE ltest SET val = val + 1 WHERE id = 1;
step s: SELECT * FROM ltest WHERE id IN (1,3) ORDER BY id;
id|val
--+---
 1|  3
 3|  2
(2 rows)

step s1c: COMMIT;

starting permutation: s1b u3 u3 s u3 s u1 s s1s u1 s s1c
step s1b: BEGIN TRANSACTION ISOLATION LEVEL REPEATABLE READ; SELECT count(*) FROM ltest;
count
-----
  182
(1 row)

step u3: UPDATE ltest SET val = val + 1 WHERE id = 3;
step u3: UPDATE ltest SET val = val + 1 WHERE id = 3;
step s: SELECT * FROM ltest WHERE id IN (1,3) ORDER BY id;
id|val
--+---
 1|  0
 3|  2
(2 rows)

step u3: UPDATE ltest SET val = val + 1 WHERE id = 3;
step s: SELECT * FROM ltest WHERE id IN (1,3) ORDER BY id;
id|val
--+---
 1|  0
 3|  3
(2 rows)

step u1: UPDATE ltest SET val = val + 1 WHERE id = 1;
step s: SELECT * FROM ltest WHERE id IN (1,3) ORDER BY id;
id|val
--+---
 1|  1
 3|  3
(2 rows)

step s1s: SELECT * FROM ltest WHERE id IN (1,3) ORDER BY id;
id|val
--+---
 1|  0
 3|  0
(2 rows)

step u1: UPDATE ltest SET val = val + 1 WHERE id = 1;
step s: SELECT * FROM ltest WHERE id IN (1,3) ORDER BY id;
id|val
--+---
 1|  2
 3|  3
(2 rows)

step s1c: COMMIT;

starting permutation: u3 u3 s u3 s u3 s u3 s
step u3: UPDATE ltest SET val = val + 1 WHERE id = 3;
step u3: UPDATE ltest SET val = val + 1 WHERE id = 3;
step s: SELECT * FROM ltest WHERE id IN (1,3) ORDER BY id;
id|val
--+---
 1|  0
 3|  2
(2 rows)

step u3: UPDATE ltest SET val = val + 1 WHERE id = 3;
step s: SELECT * FROM ltest WHERE id IN (1,3) ORDER BY id;
id|val
--+---
 1|  0
 3|  3
(2 rows)

step u3: UPDATE ltest SET val = val + 1 WHERE id = 3;
step s: SELECT * FROM ltest WHERE id IN (1,3) ORDER BY id;
id|val
--+---
 1|  0
 3|  4
(2 rows)

step u3: UPDATE ltest SET val = val + 1 WHERE id = 3;
step s: SELECT * FROM ltest WHERE id IN (1,3) ORDER BY id;
id|val
--+---
 1|  0
 3|  5
(2 rows)

